Replace log.Fatal with error return in id_alloc

joezxy opened the pull request:
This modification is accoding to "Propagate errors from storage/id_alloc.go" in TODO.md. Previously in id_alloc.go, if error happens, log.Fatal will directly close the process, better to use error handling mechanism.

There are two points in the code: 
1) Use channel to return error. Actually RaftID allocation is samilar to NodeID allocation, only one small difference is the background allocateBlock goroutine. This goroutine needs to return error if unable to get id from KV DB, so it has to use channel to return error.

2) If allocateBlock() return error, instead of return error in next Allocate() directly, maybe it is better to use the existing allocated IDs firstly, and finally return error when the channel is empty. 

BTW, The design of this project is cool, I'm going through the issues and TODO, hope to contribute more :-)

Commit:
Replace log.Fatal with error return in id_alloc

bdarnell commented on storage/id_alloc.go:
```diff
@@ -54,21 +57,30 @@ func NewIDAllocator(idKey proto.Key, db *client.KV, minID int64, blockSize int64
+			} else {
+				return id, nil
+			}
+		case err = <-ia.err:
```
If I'm reading this correctly, once an error has been returned, all future calls to Allocate will simply hang because there is nothing left to write to either the ids or err channels. If we cannot recover, then we need to save the error to return on all future Allocate calls. But an error handler that leaves the IDAllocator in a permanently broken state may be worse than a log.Fatal. I think we have to be able to retry on errors.

Commit:
fix problem of Allocate() recovery

joezxy commented:
@bdarnell , you are correct, in the previous code, if both ids channel and error channel become empty, the next Allocate() will hang.

I add retry logic into Allocate(), suppose at the moment KV DB is down, allocateBlock will put error to channel, after retries,  Allocate() will fail and return nil, but the last background allocateBlock will leave a error inside idAlloc.err channel. When KV DB is up again, Allocate() invocation will receive that error in idAlloc.err channel, then call allocateBlock to get ID, finally continue to allocate IDs.

joezxy commented:
Hi, there is one strange issue in CircleCI testing.
In the first commit, CI report shows that gossip package test fail, but racetest success. Actually i only changed id_alloc.go, id_alloc_test.go and store.go...did not touch gossip at all. Has this kind of issue happened before?
In the second commit, CI report shows that storage package test fail, it works fine in my machine, both in latest master branch and my own branch. Is it possible because of the 30s timeout setting in circle.yml? since -vmodule=multiraft=5 will print more message, may affect it also.

Commit:
rename new testcases of id_alloc.go

joezxy commented:
After rebase the modification to the latest commit of cockroachdb/master (previous commit has 2-30 commit behind), it works now. :-)

bdarnell commented:
We need to think a little more about how we want to handle errors. With this implementation, after maxRetryTimes has been reached, one call to Allocate will fail but all the rest will keep blocking. I think we should either put the allocator into a "down" state (in which all calls to Allocate will fail until a retry eventually succeeds), or just leave all the Allocate calls blocking indefinitely until something succeeds.

joezxy commented:
@bdarnell , thanks for the comments, and I'm not sure if there is some small misunderstandings about the logic, let me try to explain my idea. 

1) Channel(ids) is first-class citizen, that is if its length is not 0, the loop will continue, and branch "id = <-ia.ids" can be reached. In most of the cases, it will return one ID, but when there is a allocationTrigger, ia.allocateBlock is called to push more ID into ia.ids if success, and it will put one error into channel(err) if fail. 
2) If channel(ids) is empty, it means allocateBlock failed to get ID in background, and it will put one error into channel(err) at the time. Now the new Allocate() will goto the case "err = <-ia.err", it will start a new goroutine to get ID again and wait, if it success, channel(ids) will be filled, then will goto "case id = <-ia.ids" again and return; if it fail, a new error will be put into channel(err) and trigger case "err = <-ia.err" again. And after several retries, Allocate() will return one invalid value and error. But it's important to be noted: even Allocate() fails after several reties, the last ia.allocateBlock goroutine is running at background, and it will put error or allocated IDs in the channel.
3) Then when the next Allocate() invocation comes, it will found either channel(ids) has allocated ID or error in channel(err), so it will not be blocked, the following process is just same as above.
4) And because case statement is randomly scheduled, case "err = <-ia.err" can be reached even if "len(ia.ids) > 0", but no problem, even then retry times is greater than maxRetryTimes, case "id = <-ia.ids"  can also be reached and return the allocated ID finally.



TestAllocateErrorAndRecovery simulates the above scenario, and it works as expectation:
0) Create one IDAllocator with minID(2) and blockSize(10)
1) Call Allcoate(), after returning ID(2), now there should be 8 IDs left in the channel(ids), and one allocationTrigger in the middle.
2) Change the Allocator's Key to trigger error when fetching ID from KV DB
3) Call Allocate() 8 times, pull all the allocated IDs from channel(ids), should get ID from 3 to10. In the middle there is one async prefetching goroutine triggered by allocationTrigger, and it will put error in channel(err) because can not access KV DB now
4) Now there is nothing in channel(ids), then again call Allocate() several times, all invocations should return (-1, err).
5) Change the Allocator's Key to the correct one, to allow fetching from KV DB
6) Then start to call Allocate(), it should be able to return ID from 11 and continue to process.

joezxy committed:
remove the unnecessary loop condition "exit"

After the retry logic is added, no need to set "exit" condition when receiving error. After all,  there will not be a condition like: "ids is empty" and "retry times is more than maximum",  and at the same time, "exit" flag is false. In the previous logic, if retry times increases, "exit" will always be set to true.

joezxy commented:
After adding retry logic, the previous "exit" flag is not necessary, did a small modification on it.

bdarnell commented:
I understand how this is working; it just feels weird. Consider ten Allocate() calls coming in at once on different threads while allocations are failing. They will all eventually return errors, but only after a total of 30 retries. Furthermore, the more calls there are, the longer it will take before they start returning errors (because on average the errors will be distributed evenly across the waiting calls, so every call will go through two retries before any call gets its third). 

This behavior isn't terrible, and may be acceptable in comparison to any more complicated scheme. I'm actually wondering if we can simplify it further by removing the limit and simply retry indefinitely in Allocate(), never returning an error (this would also let us get rid of the error channel and simply loop in allocateBlock).

Commit:
put retry logic into allocateBlock(), let allocate() block to wait

joezxy commented:
@bdarnell , get your point, simpler is better, the implementation is modified as the comments, could you take another look? Thanks.

bdarnell commented:
LGTM

tschottdorf commented on storage/id_alloc.go:
```diff
@@ -27,6 +30,15 @@ import (
 const allocationTrigger = 0
 
+// IDAllocationRetryOpts sets the retry options for handling RaftID
+// allocation error
```
nit picking: Comment should end with a dot.

tschottdorf replied:
also, s/error/errors/

tschottdorf commented on storage/id_alloc.go:
```diff
@@ -56,10 +68,11 @@ func NewIDAllocator(idKey proto.Key, db *client.KV, minID int64, blockSize int64
 }
 
-// Allocate allocates a new ID from the global KV DB. If multiple
+// Allocate allocates a new ID from the global KV DB, and it will
```
I think you can leave out the everything after the comma (it refers to an unexported function).

tschottdorf commented on storage/id_alloc_test.go:
```diff
@@ -78,9 +82,91 @@ func TestIDAllocatorNegativeValue(t *testing.T) {
 	}
 }
+
+// TestNewIDAllocatorInvalidArgs checks validation logic of NewIDAllocator
```
dot

tschottdorf commented on storage/id_alloc_test.go:
```diff
@@ -78,9 +82,91 @@ func TestIDAllocatorNegativeValue(t *testing.T) {
+	store, _ := createTestStore(t)
+	allocd := make(chan int, 10)
+
+	// firstly create a valid IDAllocator to get some ID
```
dots all the way down from here as well.

tschottdorf commented:
LGTM

Commit:
modify the comments

joezxy commented:
Thanks for your comments, @tschottdorf , could you have a look again?

tschottdorf commented:
Thanks @joezxy, LGTM

spencerkimball commented on storage/id_alloc.go:
```diff
@@ -77,9 +89,14 @@ func (ia *IDAllocator) Allocate() int64 {
+	retryOpts := IDAllocationRetryOpts
+	util.RetryWithBackoff(retryOpts, func() (util.RetryStatus, error) {
+		if err := ia.db.Call(proto.Increment, proto.IncrementArgs(ia.idKey, incr), ir); err != nil {
+			return util.RetryContinue, nil
```
This should really generate warnings.

    log.Warningf("unable to allocate %d ids from %q: %s", incr, ia.idKey, err)
